src/download.rs

```
@@ -549,3 +572,176 @@ mod error {
+    fn write_path(&self) -> Result<&Path> {
+        Ok(self.as_ref())
+    }
+
+    // no-op
+    fn finalize(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+struct FileTarget {
+    path: PathBuf,
+    temp_file: Option<NamedTempFile>,
+}
+
+impl FileTarget {
+    async fn new<P: AsRef<Path>>(path: P) -> Result<Box<dyn SnapshotWriteTarget>> {
+        let path = path.as_ref();
+        Ok(Box::new(FileTarget {
+                path: path.into(),
+                temp_file: None,
+                }
+            ))
+    }
+}
+
+impl AsRef<Path> for FileTarget {
+    fn as_ref(&self) -> &Path {
+        self.path.as_ref()
+    }
+}
+
+#[async_trait]
+impl SnapshotWriteTarget for FileTarget {
+    // truncate file to desired size
+    async fn grow(&mut self, length: i64) -> Result<()> {
+        let path = self.as_ref();
+
+        // Create a temporary file and extend it to the required size.
+        let target_dir = path
+            .parent()
+            .context(error::ValidateParentDirectory { path })?;
+
+        let temp_file = NamedTempFile::new_in(target_dir)
+            .context(error::CreateTempFile { path: target_dir })?;
+
+        let temp_file_len = length;
+        let temp_file_len = u64::try_from(temp_file_len).with_context(|| error::ConvertNumber {
+            what: "temp file length",
+            number: temp_file_len.to_string(),
+            target: "u64",
+        })?;
+
+        temp_file
+            .as_file()
+            .set_len(temp_file_len)
+            .context(error::ExtendTempFile {
+                path: temp_file.as_ref(),
+                }
+            )?;
+
+        self.temp_file.replace(temp_file);
+
+        Ok(())
+    }
+
+    fn write_path(&self) -> Result<&Path> {
+        Ok(self.temp_file.as_ref().unwrap().as_ref())
+    }
+
+    // persist file to destination
+    fn finalize(&mut self) -> Result<()> {
+        let temp_file = self.temp_file.take().unwrap();
```

Should return an error here instead of `unwrap()`, like the `error::MissingTempFile` suggestion from above.

---

Thanks. I made the change.